src/deploy/functions/remoteSource.ts

```diff
@@ -0,0 +1,221 @@
+import * as fs from "fs";
+import * as path from "path";
+import { spawnSync, SpawnSyncReturns } from "child_process";
+
+import * as tmp from "tmp";
+
+import { FirebaseError, hasMessage } from "../../error";
+import { logger } from "../../logger";
+import { logLabeledBullet } from "../../utils";
+import { resolveWithin } from "../../pathUtils";
+
+export interface GitClient {
+  clone(repository: string, destination: string): SpawnSyncReturns<string>;
+  fetch(ref: string, cwd: string): SpawnSyncReturns<string>;
+  checkout(ref: string, cwd: string): SpawnSyncReturns<string>;
+}
+
+export class DefaultGitClient implements GitClient {
+  clone(repository: string, destination: string): SpawnSyncReturns<string> {
+    return spawnSync(
+      "git",
+      ["clone", "--filter=blob:none", "--no-checkout", "--depth=1", repository, destination],
+      { encoding: "utf8" },
+    );
+  }
+
+  fetch(ref: string, cwd: string): SpawnSyncReturns<string> {
+    return spawnSync("git", ["fetch", "--depth=1", "--filter=blob:none", "origin", ref], {
+      cwd,
+      encoding: "utf8",
+    });
+  }
+
+  checkout(ref: string, cwd: string): SpawnSyncReturns<string> {
+    return spawnSync("git", ["checkout", ref], { cwd, encoding: "utf8" });
+  }
+}
+
+export async function cloneRemoteSource(
+  repository: string,
+  ref: string,
+  dir?: string,
+  gitClient: GitClient = new DefaultGitClient(),
+): Promise<string> {
+  /**
+   * Clones a Git repo to a temporary directory and returns the absolute path
+   * to the source directory. Verifies that a `functions.yaml` manifest exists
+   * before returning.
+   *
+   * @param repository Remote Git URL (e.g. https://github.com/org/repo)
+   * @param ref Git ref to fetch (tag/branch/commit)
+   * @param dir Optional subdirectory within the repo to use
+   * @param gitClient Optional Git client for testing/injection
+   * @returns Absolute path to the checked‑out source directory
+   */
+  logger.debug(`Cloning remote source: ${repository}@${ref} (dir: ${dir || "."})`);
+
+  const tmpDir = tmp.dirSync({
+    prefix: "firebase-functions-remote-",
+    unsafeCleanup: true,
+  });
+
+  if (!isGitAvailable()) {
+    throw new FirebaseError(
+      "Git is required to deploy functions from a remote source. " +
+        "Please install Git from https://git-scm.com/downloads and try again.",
```

Comment:
There's a tool called giget that can fetch code without requiring `git`. Might be worth looking at how they do it https://github.com/unjs/giget?tab=readme-ov-file#features